Workaround malformed `typeof` statements in TypeForwardedToAttributes

ericstj opened the pull request:
fix #31250
replaces https://github.com/dotnet/sdk/pull/31252

This is a workaround for https://github.com/dotnet/roslyn/issues/67341

ericstj committed:
Workaround malformed `typeof` statements in TypeForwardedToAttributes

ViktorHofer reviewed (commented):
We want to undo part of this change when https://github.com/dotnet/roslyn/issues/67341 is fixed. Please add a "TODO: // Remove when X is fixed" marker and in addition do one of the following:
- Move the code that will be undone into a separate commit and don't squash when merging.
- Group the code together so that undoing it manually is straightforward and doesn't require domain knowledge.

andriipatsula commented on src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxRewriter/TypeForwardAttributeCSharpSyntaxRewriter.cs:
```diff
@@ -10,33 +12,127 @@ namespace Microsoft.DotNet.GenAPI.SyntaxRewriter
+    ///     [assembly:TypeForwardedToAttribute(typeof((T1, T2, T3))] ->
+    ///     [assembly:TypeForwardedToAttribute(typeof(System.ValueTuple<,,>))]
     /// </summary>
     public class TypeForwardAttributeCSharpSyntaxRewriter : CSharpSyntaxRewriter
```
Recommendation for improving performance (in the future):
What if we modify the CSharpFileBuilder class so that we only run the TypeForwardAttributeCSharpSyntaxRewriter on generated type forwards [CSharpFileBuilder.cs#L91](https://github.com/dotnet/sdk/blob/main/src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs#L318), instead of applying it to the entire generated file [CSharpFileBuilder.cs#L99](https://github.com/dotnet/sdk/blob/main/src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs#L99)?

ericstj replied:
I was able to remove this completely.

ericstj commented:
> We want to undo part of this change when [dotnet/roslyn#67341](https://github.com/dotnet/roslyn/issues/67341) is fixed. Please add a "TODO: // Remove when X is fixed" marker and in addition do one of the following:
> 
> * Move the code that will be undone into a separate commit and don't squash when merging.
> * Group the code together so that undoing it manually is straightforward and doesn't require domain knowledge.

It’s quite possible that both bugs being worked around in this TypeForwardAttributeCSharpSyntaxRewriter are the same root cause and removing the entire type is what we will do. @andriipatsula do we have a bug for the first issue?

andriipatsula commented:
> do we have a bug for the first issue? System.Collections.Generic.IAsyncEnumerable<,,>

If I understood correctly, you are asking about an issue in scope of which I created the `TypeForwardAttributeCSharpSyntaxRewriter` class. It was rather a feature: to generate forwarded type assembly attributes https://github.com/dotnet/sdk/pull/30556 . My initial impression was that the behavior is expected.

ericstj commented:
I don’t think it’s expected that the typeof expressions are generated with invalid syntax.  It looks like the `TypeForwardAttributeCSharpSyntaxRewriter` was created to workaround that.  I wanted to know if we reported the invalid syntax issue to Roslyn.  If not, then it is fine — I will roll it into the same issue as the one tracking ValueTuple.

andriipatsula commented:
> I wanted to know if we reported the invalid syntax issue to Roslyn

I think - no.

ericstj committed:
Remove TypeForwardAttributeCSharpSyntaxRewriter

This was only required because we were constructing `typeof` expressions
with bound generics.  Instead we should create unbound generics.

ericstj commented:
I was able to remove the rewriter completely after better understanding the problem.  We just needed to tranform the symbol we got back from `GetTypeForwards` to be an unbound generic.  I'm still leaving the roslyn issue open since I'm not sure what they want to do, but this is a much better solution.

ViktorHofer reviewed (approved):
Thanks a lot 💯
